gdnative-core/src/geom/basis.rs

```diff
@@ -29,14 +116,316 @@ impl Basis {
         }
     }
 
-    pub fn from_diagonal(p_diag: Vector3) -> Basis {
-        Basis {
-            elements: [
-                Vector3::new(p_diag.x, 0.0, 0.0),
-                Vector3::new(0.0, p_diag.y, 0.0),
-                Vector3::new(0.0, 0.0, p_diag.z),
-            ],
+    /// Inverts the matrix.
+    pub fn invert(&mut self) {
+        let [x, y, z] = self.elements;
+
+        let co = [
+            y.y * z.z - y.z * z.y,
+            y.z * z.x - y.x * z.z,
+            y.x * z.y - y.y * z.x,
+        ];
+
+        let det: f32 = x.x * co[0] + x.y * co[1] + x.z * co[2];
+        assert!(!det.approx_eq(&0.0), "Determinant was zero");
+
+        let s: f32 = 1.0 / det;
+
+        self.set_x(Vector3::new(co[0] * s, co[1] * s, co[2] * s));
+        self.set_y(Vector3::new(
+            (x.z * z.y - x.y * z.z) * s,
+            (x.x * z.z - x.z * z.x) * s,
+            (x.y * z.x - x.x * z.y) * s,
+        ));
+        self.set_z(Vector3::new(
+            (x.y * y.z - x.z * y.y) * s,
+            (x.z * y.x - x.x * y.z) * s,
+            (x.x * y.y - x.y * y.x) * s,
+        ));
+    }
+
+    /// Returns the inverse of the matrix.
+    pub fn inverted(mut self) -> Basis {
+        self.invert();
+        self
+    }
+
+    /// Transposes the matrix.
+    pub fn transpose(&mut self) {
+        std::mem::swap(&mut self.elements[0].y, &mut self.elements[1].x);
+        std::mem::swap(&mut self.elements[0].z, &mut self.elements[2].x);
+        std::mem::swap(&mut self.elements[1].z, &mut self.elements[2].y);
+    }
+
+    /// Returns the transposed version of the matrix.
+    pub fn transposed(mut self) -> Basis {
+        self.transpose();
+        self
+    }
+
+    /// Returns the determinant of the matrix.
+    pub fn determinant(&self) -> f32 {
+        let elements = self.elements;
+        return elements[0].x * (elements[1].y * elements[2].z - elements[2].y * elements[1].z)
+            - elements[1].x * (elements[0].y * elements[2].z - elements[2].y * elements[0].z)
+            + elements[2].x * (elements[0].y * elements[1].z - elements[1].y * elements[0].z);
+    }
+
+    /// Orthonormalizes the matrix
+    ///
+    /// Performs a [Gram-Schmidt orthonormalization](https://en.wikipedia.org/wiki/Gram-Schmidt_process) on the basis of the matrix.
+    /// This can be useful to call from time to time to avoid rounding error for orthogonal matrices.
```

Comment:
Typo: `rounding error` -> `rounding errors`.

Doc-comment should have a `Panic` section.